### DISCUSSION TITLE
How can I assign a key or key combination to the Hand Tool?

### DISCUSSION THREAD

**Discussion by vladQll** [xournalpp/xournalpp#5039] [Q&A] [answered]
Since I have used up all the buttons on my tablet and there is no key assigned to the hand tool, I need to know how I can assign a key to it. I have practiced a lot in some graphic design software so much that I used to press the space key to use the hand tool, I want to imitate that action by assigning the space key to the hand tool in Xournal++.

**Comment by netlimpopo** [comment]
Keybindings require the transition from Gtk3 to Gtk4. Some developers are already working on this topic. 
As of right now, individual shortcuts are not possible.  
The hand tool tool can be activated by ctrl + shift + A. Alternatively you can assign a standard tool as the hand tool and use crtl + +Shift+ default/  ctrl + shift + d. 
Other than that I am afraid, no you can't.  

See also #919 .

**Comment by rolandlo** [comment]
@vladQll You can write a simple [plugin](https://xournalpp.github.io/guide/plugins/plugins/) to do exactly what you want.

**Reply by vladQll** [reply]
hello @rolandlo yes I was thinking about it recently and I remembered that I had modified a shortcutplugin that you left some time ago in https://github.com/xournalpp/xournalpp/issues/919#issuecomment-700247463, so I have been configuring several things and it is already looking like I would like, I would like to know what is the name of the variable or constant name of the hand tool and how I could add it to a function, right now I am using the second button of the stylus for the selection tools I think there could go the hand tool in the initial position, my doubt is if it will only be necessary to add the name in quotes of the hand tool, or if you should add something more in the function here I send a visual map of the code
![image](https://github.com/xournalpp/xournalpp/assets/28522011/2d101279-f7d2-4f4f-b81f-47790b0a5bb9)

**Reply by rolandlo** [reply, marked as answer]
The constant is `'ACTION_TOOL_HAND'`. For the selection tool, you currently start with `'ACTION_TOOL_SELECT_'`, so you can't simply add `HAND` to `selectList`.
But you could replace `'ACTION_TOOL_SELECT_'` by `'ACTION_TOOL_'` and write
``` lua
selectList = {'SELECT_RECT', 'SELECT_REGION', 'SELECT_OBJECT', 'HAND'}
```

**Reply by vladQll** [reply]
How about if it worked for me just one last thing I hope not to bother you @rolandlo , what should I add so that every time I call the cycle I start with the first tool , because for example if I am writing and I call the cycle and choose select region and go back to the pencil and then I call the cycle again but this time to use the hand tool , but when I press the button that calls the cycle the cycle is placed and goes to the next tool , and I would have to press 2 or 3 more times to get to the hand tool, this is not a primary inconvenience at the moment, but suppose the tool list is long and at the moment I am mostly using the first tool it would be impractical to have to go through all the tools to get to the first tool.

**Reply by rolandlo** [reply]
With 
```lua
local type = app.getToolInfo("active")["type"]
``` 
you could get the type of the active tool. If it's neither of 
`"selectRect"`, `"selectRegion"`, `"selectObject"`, `"hand"` you set `currentSelect` to `1`.

So modify the function `select` like this:

```lua
function select()
  local type = app.getToolInfo("active")["type"]
  if type == "selectRect" or type == "selectRegion" or type == "selectObject" or type == "hand" then
    currentSelect = currentSelect % #selectList + 1
  else
    currentSelect = 1
  end
  app.uiAction({["action"] = 'ACTION_TOOL_' .. selectList[currentSelect]})
  print('ACTION_TOOL_' .. selectList[currentSelect])
end
``` 

In case you add more tools to the list, check https://github.com/xournalpp/xournalpp/blob/51694484f551a8c72085d69aec0b481d70c47813/src/core/control/ToolEnums.cpp#L111-L162 for the names.

**Reply by vladQll** [reply]
hello @rolandlo  , it seems to be in conflict because after adding the code in the select function, my other functions stopped working and did not run, I attach some of the screenshots of the lists and functions that I have

lists
![image](https://github.com/xournalpp/xournalpp/assets/28522011/92588ddb-8652-49a9-a69e-524124dfd8d0)

functions
![image](https://github.com/xournalpp/xournalpp/assets/28522011/71666cab-9e27-41e4-9785-e68862cab8b7)

**Reply by rolandlo** [reply]
There is a typo in the `if`-statement (which I have corrected above). It should end with `then` and not with `do`.

**Reply by vladQll** [reply]
thank you very much @rolandlo, now if it works, thank you very much, greetings from afar.
